services/backup/backup_service.go

```diff
@@ -371,3 +470,172 @@ func (s *Service) prepareBackupJob(
 
 	return res, dbConfig, nil
 }
+
+func vendorToServiceType(vendor string) (models.ServiceType, error) {
+	serviceType := models.ServiceType(vendor)
+	switch serviceType {
+	case models.MySQLServiceType,
+		models.MongoDBServiceType:
+	case models.PostgreSQLServiceType,
+		models.ProxySQLServiceType,
+		models.HAProxyServiceType,
+		models.ExternalServiceType:
+		return "", status.Errorf(codes.Unimplemented, "unimplemented service: %s", serviceType)
+	default:
+		return "", status.Errorf(codes.Internal, "unknown service: %s", serviceType)
+	}
+
+	return serviceType, nil
+}
+
+func (s *Service) findArtifactCompatibleServices(
+	q *reform.Querier,
+	serviceID string,
+	serviceType models.ServiceType,
+	artifactDBVersion string,
+) ([]*models.Service, error) {
+	// allow restore to the same service if db version is unknown or service type is MongoDB.
+	if artifactDBVersion == "" || serviceType == models.MongoDBServiceType {
+		service, err := models.FindServiceByID(q, serviceID)
+		if err != nil {
+			s.l.WithError(err).Warnf("restore is not possible to the same service id %q", serviceID)
+			return []*models.Service{}, nil
+		}
+
+		return []*models.Service{service}, nil
+	}
+
+	filter := models.FindServicesSoftwareVersionsFilter{ServiceType: &serviceType}
+	svs, err := models.FindServicesSoftwareVersions(q, filter, models.SoftwareVersionsOrderByServiceID)
+	if err != nil {
+		return nil, err
+	}
+
+	compatibleServiceIDs := make([]string, 0, len(svs))
+	for _, sv := range svs {
+		svm := softwareVersionsMap(sv.SoftwareVersions)
+		if err := mySQLSoftwaresInstalledAndCompatible(svm); err != nil {
+			s.l.WithError(err).Debugf("skip incompatible service id %q", sv.ServiceID)
+			continue
+		}
+
+		serviceDBVersion := svm[models.MysqldSoftwareName]
+		if artifactDBVersion != serviceDBVersion {
+			s.l.Debugf("skip incompatible service id %q: artifact version %q != db version %q\"", sv.ServiceID,
+				artifactDBVersion, serviceDBVersion,
+			)
+			continue
+		}
```

Comment:
so we check compatibility of xtradbbackup with mysql, right? Not compatibility of different versions of MySQL?

Reply:
In that case I would recommend to move this equality check above.

Reply:
done